File: frameworks/go/chi/UnitTests/errors_test.go
package unittests

import (
	"net/http"
	"testing"
)

// rb:test errors.unmatched
func TestAPathNoRouteMatchesIsNetHTTPs404(t *testing.T) {
	t.Run("errors.unmatched", func(t *testing.T) {
		response := get(t, "/errors/unmatched")

		assertStatus(t, response, http.StatusNotFound)
		if body := string(bodyOf(t, response)); body != "404 page not found\n" {
			t.Fatalf("body %q", body)
		}
	})
}

// rb:test errors.wrong_method
func TestAMethodThePathHasNoRouteForIs405WithTheMethodsItHas(t *testing.T) {
	t.Run("errors.wrong_method", func(t *testing.T) {
		response := send(t, http.MethodPost, "/items/17", nil)

		assertStatus(t, response, http.StatusMethodNotAllowed)
		if allow := response.Header.Values("Allow"); len(allow) != 5 {
			t.Fatalf("allow %v", allow)
		}
		if body := bodyOf(t, response); len(body) != 0 {
			t.Fatalf("body %q", body)
		}
	})
}

// rb:test errors.not_found
func TestAMissingRowIsTheExamplesNotFound(t *testing.T) {
	t.Run("errors.not_found", func(t *testing.T) {
		response := get(t, "/items/999999")

		assertStatus(t, response, http.StatusNotFound)
		assertJSON(t, map[string]any{"status": "Resource not found."}, response)
	})
}

// rb:test errors.malformed
func TestABodyThatIsNotJSONIsRefusedWithTheDecodersMessage(t *testing.T) {
	t.Run("errors.malformed", func(t *testing.T) {
		response := postJSON(t, http.MethodPost, "/body/validate/small", []byte(`{"customerId": 1, "lines": [`))

		assertStatus(t, response, http.StatusBadRequest)
		assertJSON(t, map[string]any{"status": "Invalid request.", "error": "unexpected EOF"}, response)
	})
}

File: frameworks/node/h3/UnitTests/errors.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { app } from "./app.ts";

/**
 * errors: every refusal is h3's own error JSON, written by its error handling for the router's miss,
 * for readBody's failure and for the HTTPError the items handlers throw. Nothing here reshapes any
 * of them.
 */

// rb:test errors.unmatched
test("errors.unmatched: a path no route matches is the router's 404, naming the request's URL", async () => {
  const response = await app.request("/errors/unmatched");

  assert.equal(response.status, 404);
  assert.deepEqual(await response.json(), { status: 404, message: "Cannot find any route matching [GET] http://localhost/errors/unmatched" });
});

// rb:test errors.not_found
test("errors.not_found: an id with no row is the handler's HTTPError 404", async () => {
  const response = await app.request("/items/999999");

  assert.equal(response.status, 404);
  assert.deepEqual(await response.json(), { status: 404, statusText: "Not Found", message: "Not Found" });
});

// rb:test errors.wrong_method
test("errors.wrong_method: a method the path has no route for is 404, as a path with no route is", async () => {
  const response = await app.request("/items/17", { method: "POST" });

  assert.equal(response.status, 404);
  assert.equal(((await response.json()) as { message: string }).message, "Cannot find any route matching [POST] http://localhost/items/17");
});

// rb:test errors.malformed
test("errors.malformed: a body that is not JSON is readBody's 400, before any schema runs", async () => {
  const response = await app.request("/body/validate/small", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: '{"customerId": 1, "lines": [',
  });

  assert.equal(response.status, 400);
  assert.deepEqual(await response.json(), { status: 400, statusText: "Bad Request", message: "Invalid JSON body" });
});
